### PR TITLE
Core Trainer Refactors

### PR DESCRIPTION
**Pull request by ASzot**
## Motivation and Context

Changes required for SIRo merge.

## How Has This Been Tested

End-to-end training.

## Types of changes

- **\[Refactoring\]** Large changes to the code that improve its functionality or performance
- **\[Development\]** A pull request that add new features to the [habitat-lab](/habitat-lab) task and environment codebase. Development Pull Requests must be small (less that 500 lines of code change), have unit testing, very extensive documentation and examples. These are typically new tasks, environments, sensors, etc... The review process for these Pull Request is longer because these changes will be maintained by our core team of developers, so make sure your changes are easy to understand!

### CODE REVIEW

File: habitat-lab/habitat/datasets/rearrange/rearrange_generator.py [hunk 1]
```diff
@@ -825,6 +825,8 @@ def initialize_sim(self, scene_name: str, dataset_path: str) -> None:
             for object_path in self.cfg.additional_object_paths:
                 object_attr_mgr.load_configs(osp.abspath(object_path))
         else:
+            if self.sim.config.sim_cfg.scene_id != scene_name:
```
Comment (rpartsey) [inline habitat-lab/habitat/datasets/rearrange/rearrange_generator.py:828]: These two if statements are mutually exclusive. We can use if else here.

**Review by rpartsey — APPROVED** [review]
Overall LGTM, thought I haven't tried running training/eval commands (I don't know what are the up-to-date commands).

If training / evaluation is working, I'm ok with merging it.

File: habitat-baselines/habitat_baselines/rl/ppo/ppo_trainer.py [hunk 2]
```diff
@@ -852,302 +842,35 @@ def _eval_checkpoint(
...
 
         self._agent = self._create_agent(None)
-        action_shape, discrete_actions = get_action_space_info(
-            self._agent.policy_action_space
-        )
-
         if self._agent.actor_critic.should_load_agent_state:
             self._agent.load_state_dict(ckpt_dict)
 
-        observations = self.envs.reset()
```
Comment (xavierpuigf) [inline habitat-baselines/habitat_baselines/rl/ppo/ppo_trainer.py:862]: Is all this going to test_baselines_trainer?

File: habitat-baselines/habitat_baselines/rl/ddppo/policy/resnet_policy.py [hunk 3]
```diff
@@ -134,6 +134,7 @@ def from_config(
                 )
             )
         )
+
```
Comment (xavierpuigf) [inline habitat-baselines/habitat_baselines/rl/ddppo/policy/resnet_policy.py:137]: can be removed?

File: habitat-baselines/habitat_baselines/rl/hrl/hrl_rollout_storage.py [hunk 4]
```diff
@@ -96,9 +98,9 @@ def insert(
```
Comment (xavierpuigf) [inline habitat-baselines/habitat_baselines/rl/hrl/hrl_rollout_storage.py:98]: Add here 
```
self._last_should_inserts = should_inserts
```
Before returning

Comment (xavierpuigf) [reply habitat-baselines/habitat_baselines/rl/hrl/hrl_rollout_storage.py:98]: See this PR
https://github.com/facebookresearch/habitat-lab/commit/ea2845e132f58e7769b956464ea28de1235a9eee

File: habitat-baselines/habitat_baselines/rl/hrl/hrl_rollout_storage.py [hunk 5]
```diff
@@ -196,8 +186,8 @@ def data_generator(self, advantages, num_batches) -> Iterator[DictTree]:
                 # Stricly less than is so we throw out the last transition. We
                 # need to throw out the last transition because we were not
                 # able to accumulate rewards for it.
-                batch["loss_mask"][:, i] = (
-                    batch["loss_mask"][:, i] < self._cur_step_idxs[env_i]
+                batch["loss_mask"][:, i] = batch["loss_mask"][:, i] < (
+                    self._cur_step_idxs[env_i] - 1
                 )
```
Comment (xavierpuigf) [inline habitat-baselines/habitat_baselines/rl/hrl/hrl_rollout_storage.py:191]: There used to be this. We dont need it anymore?

```
            batch["rnn_build_seq_info"] = build_rnn_build_seq_info(
                device=self.device,
                build_fn_result=build_pack_info_from_dones(
                    dones_cpu[0 : self.num_steps, inds.numpy()].reshape(
                        -1, len(inds)
                    ),
                ),
            )

```

**Review by xavierpuigf — COMMENTED** [review]
Looks great! I left a few comments, the most important is in hrl_rollout_storage, where a line is missing that caused a painful bug in the past. Otherwise ready to merge.

**Pull request merged**